src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexCompiler.cs

```diff
@@ -3925,6 +3940,121 @@ void EmitStackPop()
             }
         }
 
+        protected void EmitScan(DynamicMethod findFirstCharMethod, DynamicMethod goMethod)
+        {
+            LocalBuilder bump = DeclareInt32();
+            LocalBuilder stoppos = DeclareInt32();
+            Label returnLabel = DefineLabel();
+
+            Label notRightToLeft = DefineLabel();
+            // int bump = 1
+            Ldc(1);
+            Stloc(bump);
+
+            // int stoppos = text.Length
+            _ilg!.Emit(OpCodes.Ldarga_S, 1);
+            Call(s_spanGetLengthMethod);
+            Stloc(stoppos);
+
+            // if (runregex.RightToLeft)
+            // {
+            Ldthisfld(s_runregexField);
+            Callvirt(s_regexGetRightToLeft);
+            BrfalseFar(notRightToLeft);
+
+            // bump = -1;
+            // stoppos = 0;
+            // }
+            Ldc(-1);
+            Stloc(bump);
+            Ldc(0);
+            Stloc(stoppos);
+            MarkLabel(notRightToLeft);
+
+            // while (true)
+            Label whileLoopEnd = DefineLabel();
+            Label whileLoopBody = DefineLabel();
+            MarkLabel(whileLoopBody);
+
+            // if (FindFirstChar(text))
+            Label afterFindFirstCharLabel = DefineLabel();
+            Ldthis();
+            Ldarg_1();
+            Call(findFirstCharMethod);
+            BrfalseFar(afterFindFirstCharLabel);
+
+            // CheckTimeout();
+            Ldthis();
+            Call(s_checkTimeoutMethod);
+
+            // Go(text);
+            Ldthis();
+            Ldarg_1();
+            Call(goMethod);
+
+            // if (runmatch!._matchcount[0] > 0)
+            //   return;
+            Label afterSuccessMatchLabel = DefineLabel();
+            Ldthisfld(s_runmatchField);
+            _ilg!.Emit(OpCodes.Ldfld, s_matchMatchCountField);
+            Ldc(0);
+            LdelemI4();
+            Ldc(0);
+            _ilg!.Emit(OpCodes.Cgt);
+            BrfalseFar(afterSuccessMatchLabel);
+            BrFar(returnLabel);
```

Comment:
We don't want to do the same thing we do in source gen and have Go return a bool rather than having this code reach into the match object to inspect the match count?  The bool should be faster and simpler.

Reply:
We can do that as well here. I originally did it only in the source generator because main idea with that is that source generator doesn't have access to `match._matchCount[0]` field, so I used `Go` return value to let me know if the input matched or not. For Compiled engine on the other hand, we do have access to the _matchCount field which is why I left it as is, but I agree it would be better to match for a) consistency between generated engines and b) perf.